[clang-tidy] Add new check 'misc-scope-reduction' (#175429)
llvm/llvm-project: clang-tools-extra/docs/clang-tidy/checks/misc/scope-reduction.rst

.. title:: clang-tidy - misc-scope-reduction

misc-scope-reduction

> ===========================

Reviewer: ```suggestion
====================
```

Detects local variables in functions whose scopes can be minimized. This check
covers guidelines described by SEI DCL19-C, MISRA C++:2008 Rule 3-4-1, and MISRA
C:2012 Rule 8-9.

Examples:

.. code-block:: cpp

    void test_deep_nesting() {
      int deep = 1; // 'deep' can be declared in a smaller scope
      if (true) {
        if (true) {
          if (true) {
            if (true) {
              int result = deep * 4;
            }
          }
        }
      }
    }

    void test_switch_multiple_cases(int value) {
      int accumulator = 0; // 'accumulator' can be declared in a smaller scope
      switch (value) {
        case 1:
          accumulator += 10;
          break;
        case 2:
          accumulator += 20;
          break;
      }
    }

    void test_for_loop_expressions() {
      int i; // 'i' can be declared in the for-loop initialization
      for (i = 0; i < 5; i++) {
        // loop body
      }
    }

References

> ----------

Reviewer: ```suggestion
----------

```

.. title:: clang-tidy - misc-scope-reduction

misc-scope-reduction
====================

Detects local variables in functions whose scopes can be minimized.

Examples:

.. code-block:: cpp

    void test_deep_nesting() {
      int deep = 1; // 'deep' can be declared in a smaller scope
      if (true) {
        if (true) {
          if (true) {
            if (true) {
              int result = deep * 4;
            }
          }
        }
      }
    }

    void test_switch_case(int value) {
      int result = 0; // 'result' can be declared in a smaller scope
      switch (value) {
        case 1:
          result = 10;
          break;
        default:
          break;
      }
    }

    void test_for_loop_expressions() {
      int i; // 'i' can be declared in the for-loop initialization
      for (i = 0; i < 5; i++) {
        // loop body
      }
    }

Limitations
-----------

> This checker cannot currently detect when a variable's previous value affects

Reviewer 2: ```suggestion
This check cannot currently detect when a variable's previous value affects
```

subsequent iterations, resulting in false positives in some cases. This can
be addressed by implementing a pattern matcher that recognizes this

> accumulator pattern across loop iterations or by using clang's builtin

Reviewer 2: ```suggestion
accumulator pattern across loop iterations or by using Clang's built-in
```
